Add coloredlogs and selective trace loggers

Numerlor opened the pull request:
## Relevant Issues
closes #854, closes #706


## Description
Adds coloredlogs to the bot and standardizes the logging format with bot. I've also moved the log files to a logs directory in the project root and renamed the file name from `hackbot.log` to `sir-lancebot.log`. 
Also adds selective trace loggers from python-discord/bot#1529

The trace loggers env var should be documented in the contributing guidelines to be reachable from lancebot's guide

brad90four commented on bot/log.py:
```diff
@@ -0,0 +1,96 @@
+            **coloredlogs.DEFAULT_LEVEL_STYLES,
+            "trace": {"color": 246},
+            "critical": {"background": "red"},
+            "debug": coloredlogs.DEFAULT_LEVEL_STYLES["info"]
```
``

Numerlor replied:
resolved through a fixup on 6d7f46fabe530e4cafdffa25d7fced730965e481

brad90four reviewed (approved):
Most of this looks like black magic to me, but I don't see anything obviously bad. The one thing I did see is "missing" a trailing comma.

Bluenix2 commented on bot/log.py:
```diff
@@ -0,0 +1,96 @@
+
+        elif level_filter.startswith("!"):
+            logging.getLogger().setLevel(logging.TRACE)
+            for logger_name in level_filter.strip("!,").split(","):
+                logging.getLogger(logger_name).setLevel(logging.DEBUG)
```
Shouldn't there be another loop that sets all of them to TRACE?

Numerlor replied:
The root logger is set to TRACE so loggers we don't set to DEBUG explicitly inherit that

Bluenix2 commented:
Is there anything I need to do to test this? I have both rebuilt the container with `docker-compose build` and removed it using `docker-compose rm`.

I added this logging to make sure changes were applied:
```python
def setup() -> None:
    ...  # Setup of all the loggers

    coloredlogs.install(level=logging.TRACE, logger=root_logger, stream=sys.stdout)
    root_logger.info('Installed coloredlogs')
```

Yet still there's no colours like the other bot has, you can see that Docker adds colors so they should display in my terminal no?

![image](https://user-images.githubusercontent.com/38372706/140981530-611dc571-e496-4728-b0f8-aed4ef7781de.png)

I am on Windows 10, using what I assume is Powershell through Visual Studio Code.

Numerlor commented:
> Is there anything I need to do to test this? I have both rebuilt the container with `docker-compose build` and removed it using `docker-compose rm`.
> 
> I added this logging to make sure changes were applied:
> 
> ```python
> def setup() -> None:
>     ...  # Setup of all the loggers
> 
>     coloredlogs.install(level=logging.TRACE, logger=root_logger, stream=sys.stdout)
>     root_logger.info('Installed coloredlogs')
> ```
> 
> Yet still there's no colours like the other bot has, you can see that Docker adds colors so they should display in my terminal no?

Ah, I forgot to test Docker; it looks like the colours aren't applied because it's not tty when running through Docker.
Do we want to do enable it through `tty: true` in docker-compose like bot does or by simply forcing coloredlogs to colour output with `isattay=True` in the install call?

Bluenix2 commented:
> > Is there anything I need to do to test this? I have both rebuilt the container with `docker-compose build` and removed it using `docker-compose rm`.
> > I added this logging to make sure changes were applied:
> > ```python
> > def setup() -> None:
> >     ...  # Setup of all the loggers
> > 
> >     coloredlogs.install(level=logging.TRACE, logger=root_logger, stream=sys.stdout)
> >     root_logger.info('Installed coloredlogs')
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Yet still there's no colours like the other bot has, you can see that Docker adds colors so they should display in my terminal no?
> 
> Ah, I forgot to test Docker; it looks like the colours aren't applied because it's not tty when running through Docker. Do we want to do enable it through `tty: true` in docker-compose like bot does or by simply forcing coloredlogs to colour output with `isattay=True` in the install call?

I think we should just follow what bot does

Numerlor commented:
> I think we should just follow what bot does

Should work now

jchristgit reviewed (approved):
Looks good to me. From my understanding, this is only enabled if a tty is detected, so no weird formatting errors should crop up.

Xithrius reviewed (approved):
I see good stuff
